### ISSUE TITLE
kubeapi server ignores manifest changes

### ISSUE THREAD

**Issue by jseparovic** [kubernetes/kubernetes#112861]
### What happened?

After updating /etc/kubernetes/manifests/kube-apiserver.yaml and adding --feature-gates=MixedProtocolLBService=true to the command array, the apiserver container restarts as expected. However, upon restart, it does not update the command when inspecting with kubectl get pod -o=yaml or ps -ef.
I do notice that if I change the name of the pod in the manifest, then this new pod does have the updated command, but the old pod is not removed, so they both run and conflict with each other.

### What did you expect to happen?

I expected the apiserver to restart and the new command arg to be present in the pod output and ps -ef output.

### How can we reproduce it (as minimally and precisely as possible)?
```
[root@master1.kube3 ~]$ cat /etc/kubernetes/manifests/kube-apiserver.yaml | grep MixedProtocolLBService
    - --feature-gates=MixedProtocolLBService=true
[root@master1.kube3 ~]$ ksg pod/kube-apiserver-master1.kube3 -o=yaml | grep MixedProtocolLBService
[root@master1.kube3 ~]$ ksg pod/kube-apiserver-master1.kube3
NAME                                      READY   STATUS    RESTARTS   AGE
kube-apiserver-master1.kube3   1/1     Running   6          48d
[root@master1.kube3 ~]$ ksx pod kube-apiserver-master1.kube3
pod "kube-apiserver-master1.kube3" deleted
[root@master1.kube3 ~]$ ksg pod/kube-apiserver-master1.kube3
NAME                                      READY   STATUS    RESTARTS   AGE
kube-apiserver-master1.kube3   1/1     Running   6          8s
[root@master1.kube3 ~]$ ksg pod/kube-apiserver-master1.kube3 -o=yaml | grep MixedProtocolLBService
[root@master1.kube3 ~]$ cat /etc/kubernetes/manifests/kube-apiserver.yaml | grep MixedProtocolLBService
    - --feature-gates=MixedProtocolLBService=true
[root@master1.kube3 ~]$
```

### Anything else we need to know?

```
cat /etc/kubernetes/manifests/kube-apiserver.yaml
...
  - command:
    - kube-apiserver
    - --service-node-port-range=80-32767
    - --advertise-address=<omitted>
    - --allow-privileged=true
    - --authorization-mode=Node,RBAC
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --enable-admission-plugins=NodeRestriction
    - --enable-bootstrap-token-auth=true
    - --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt
    - --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt
    - --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key
    - --etcd-servers=https://127.0.0.1:2379
    - --feature-gates=MixedProtocolLBService=true
    - --insecure-port=0
    - --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt
    - --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key
    - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
    - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
    - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --requestheader-extra-headers-prefix=X-Remote-Extra-
    - --requestheader-group-headers=X-Remote-Group
    - --requestheader-username-headers=X-Remote-User
    - --secure-port=6443
    - --service-account-issuer=https://kubernetes.default.svc.cluster.local
    - --service-account-key-file=/etc/kubernetes/pki/sa.pub
    - --service-account-signing-key-file=/etc/kubernetes/pki/sa.key
    - --service-cluster-ip-range=10.96.0.0/12
    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
...
```    

```
kubectl -n kube-system get pod/kube-apiserver-master1.kube3 -o=yaml
...
spec:
  containers:
  - command:
    - kube-apiserver
    - --advertise-address=<omitted>
    - --allow-privileged=true
    - --authorization-mode=Node,RBAC
    - --client-ca-file=/etc/kubernetes/pki/ca.crt
    - --enable-admission-plugins=NodeRestriction
    - --enable-bootstrap-token-auth=true
    - --etcd-cafile=/etc/kubernetes/pki/etcd/ca.crt
    - --etcd-certfile=/etc/kubernetes/pki/apiserver-etcd-client.crt
    - --etcd-keyfile=/etc/kubernetes/pki/apiserver-etcd-client.key
    - --etcd-servers=https://127.0.0.1:2379
    - --insecure-port=0
    - --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt
    - --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key
    - --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
    - --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt
    - --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key
    - --requestheader-allowed-names=front-proxy-client
    - --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt
    - --requestheader-extra-headers-prefix=X-Remote-Extra-
    - --requestheader-group-headers=X-Remote-Group
    - --requestheader-username-headers=X-Remote-User
    - --secure-port=6443
    - --service-account-issuer=https://kubernetes.default.svc.cluster.local
    - --service-account-key-file=/etc/kubernetes/pki/sa.pub
    - --service-account-signing-key-file=/etc/kubernetes/pki/sa.key
    - --service-cluster-ip-range=10.96.0.0/12
    - --tls-cert-file=/etc/kubernetes/pki/apiserver.crt
    - --tls-private-key-file=/etc/kubernetes/pki/apiserver.key
...
```

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.14", GitCommit:"0f77da5bd4809927e15d1658fb4aa8f13ad890a5", GitTreeState:"clean", BuildDate:"2022-06-15T14:11:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Baremetal/Kubelet
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
AlmaLinux 8.6 (Sky Tiger)
$ uname -a
# paste output here
4.18.0-372.19.1.el8_6.x86_64
# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>
containerd
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
weave
</details>

**Comment by jseparovic** [comment]
Just to note on a workaround I'm able to do to get it configured:
- Rename Pod name in manifest to kube-apiserver2
- Kill the original apiserver process so the new one kicks in
- Configure LoadBalancer service with mixed protocols
- Rename Pod back to kube-apiserver

This allows me to start kube-apiserver temporarily with MixedProtocolLBService=true so I can create the loadbalancer.

**Comment by BenTheElder** [comment]
If I'm reading this correctly this is about the api server static pod manifest in kubeadm https://github.com/kubernetes/kubeadm/issues/2763
/sig node

**Comment by SergeyKanzhelev** [comment]
1.21 is out of support, can you try newer version of Kubernetes in case the issue was already fixed.

/assign @rphillips 

/triage needs-information

**Comment by jseparovic** [comment]
Is 1.22 in support? I’m using
https://github.com/planetscale/vitess-operator and that’s their highest
supported version

**Comment by SergeyKanzhelev** [comment]
> Is 1.22 in support? I’m using https://github.com/planetscale/vitess-operator and that’s their highest supported version

Yes, 1.22 is supported https://kubernetes.io/releases/ We encourage to test things on latest versions to simplify our triage process and ensure that we spend time effectively.

During triage @rphillips mentioned it may be by design that the command line argument update is not supported. I found this documentation https://kubernetes.io/docs/concepts/workloads/pods/#pod-update-and-replacement but it doesn't specify that this affects static pods as well. Perhaps @rphillips can confirm.

You can always delete and recreate as a workaround: https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/#dynamic-addition-and-removal-of-static-pods

**Comment by BenTheElder** [comment]
FWIW: If you know you need additional apiserver flags when creating the cluster you can add extra arguments via kubeadm config.

https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/


```yaml
apiVersion: kubeadm.k8s.io/v1beta3
kind: ClusterConfiguration
apiServer:
  extraArgs:
    feature-gates: MixedProtocolLBService=true
 ```

**Comment by jseparovic** [comment]
> You can always delete and recreate as a workaround: https://kubernetes.io/docs/tasks/configure-pod-container/static-pod/#dynamic-addition-and-removal-of-static-pods

Perfect! This process fixed the issue for me in 1.21.
